### ISSUE TITLE
discuss: When import routes, whether duplicate routes should be allowed to be imported

### ISSUE THREAD

**Issue by nic-chen** [apache/apisix-dashboard#1468]
The current practice is to directly reject duplicate routes

The reason for this is that users may import many routes at once.
It may even import the same file multiple times due to misoperation
This will result in a lot of useless data.

But on the other hand, manually creating routes can create duplicate routes.

Welcome discuss.

**Comment by nic-chen** [comment]
cc @liuxiran  @imjoey @Jaycean @juzhiyuan @membphis

**Comment by membphis** [comment]
How to know that the two imported routes are the same? Is there a unique field like ID?

**Comment by starsz** [comment]
> How to know that the two imported routes are the same? Is there a unique field like ID?

Maybe we need to compare the field and value.

**Comment by Jaycean** [comment]
Yes, when I finished importing and exporting, I discussed this problem with @liuxiran @imjoey 

Apisix allows users to create multiple duplicate routes, which adds more flexibility.

If the import does not remove the duplicate data, it is necessary to add the unique identification of the route (it does not have to be an ID, as long as it can identify the uniqueness, and if the uniqueness is repeated, it can be de duplicated), so that the dirty data of the imported route can not be generated.

**Comment by nic-chen** [comment]
@membphis @Jaycean 

I think the routing conditions are exactly the same, it can be regarded as duplicate. Condition fields include uri, uris, host, hosts, vars, remote_addr, remote_addrs and priority etc.

**Comment by membphis** [comment]
> I think the routing conditions are exactly the same, it can be regarded as duplicate. Condition fields include uri, uris, host, hosts, vars, remote_addr, remote_addrs and priority etc.

It is not a good way.

> it does not have to be an ID, as long as it can identify the uniqueness, and if the uniqueness is repeated, it can be de duplicated

this way is better

**Comment by liuxiran** [comment]
How about leave the choice to users? which means when users import route, they can choose
* `duplicate is allowed`: that means any routes in right configuration can be imported, no matter it is duplicate or not, as APISIX allows to create the same route, IMHO, this operation is reasonable
* `Cover the duplicate data`: that means if the route to be imported is duplicate with the one already in APISIX, we will use the latest one to cover the old one. and we can set this option as a default one. Then what is a duplicate route? Maybe we can circle a minimum set

**Comment by nic-chen** [comment]
> * duplicate is allowed

`duplicate is allowed` is good. but I'm not sure about `Cover the duplicate data`, there are certain risks in doing so.

**Comment by membphis** [comment]
> duplicate is allowed
> Cover the duplicate data

they are both not good.

Are there any fields like ID to identify a specific object (such as route, upstream, etc.)?

**Comment by nic-chen** [comment]
> Are there any fields like ID to identify a specific object (such as route, upstream, etc.)? 

the data exported from APISIX have ID for routes, but data generated from code or other sources may not have

**Comment by starsz** [comment]
`duplicate is allowed`

I think it's good. Because the duplicate route is harmless, we can keep it.

**Comment by batman-ezio** [comment]
I see this PR https://github.com/apache/apisix-dashboard/pull/1606/files
so now `duplicate is not allowed`?

**Comment by membphis** [comment]
> I think it's good. Because the duplicate route is harmless, we can keep it.

When the user needs to update the route, this duplicate route also needs to be updated.

If the duplicate route is not updated, the new route will not take effect and cause bugs.

**Comment by starsz** [comment]
> I see this PR https://github.com/apache/apisix-dashboard/pull/1606/files
> so now `duplicate is not allowed`?

This issue is for the situation when we import routes.


> When the user needs to update the route, this duplicate route also needs to be updated.
If the duplicate route is not updated, the new route will not take effect and cause bugs.

Why the duplicate also needs to be updated? IMO, if the duplicate route is not updated, they are two different routes.

**Comment by membphis** [comment]
> Why the duplicate also needs to be updated? IMO, if the duplicate route is not updated, they are two different routes.

that is why I think it'll cause a but. because there are two different routes.

**Comment by membphis** [comment]
I think everyone is right, sorry for causing you trouble. 

https://github.com/apache/apisix-dashboard/issues/1468#issuecomment-781184877 is a good way.
